File: ios/ConnectivityMonitor.swift
import Foundation
import Network

/// iOS connectivity monitor using NWPathMonitor
/// Tracks network status and notifies the C++ engine selector
class ConnectivityMonitor: ObservableObject {
    @Published var isOnline: Bool = false
    @Published var connectionType: ConnectionType = .unknown
    
    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "ConnectivityMonitor")
    
    // Callback to C++ engine selector
    private var statusCallback: ((Bool) -> Void)?
    
    enum ConnectionType {
        case unknown
        case wifi
        case cellular
        case ethernet
        case other
    }
    
    init() {
        startMonitoring()
    }
    
    deinit {
        monitor.cancel()
    }
    
    /// Start monitoring network connectivity
    func startMonitoring() {
        monitor.pathUpdateHandler = { [weak self] path in
            DispatchQueue.main.async {
                self?.updateConnectionStatus(path: path)
            }
        }
        monitor.start(queue: queue)
    }
    
    /// Stop monitoring network connectivity
    func stopMonitoring() {
        monitor.cancel()
    }
    
    /// Set callback for connectivity status changes
    func setStatusCallback(_ callback: @escaping (Bool) -> Void) {
        statusCallback = callback
    }
    
    /// Update connection status based on network path
    private func updateConnectionStatus(path: NWPath) {
        let wasOnline = isOnline
        isOnline = path.status == .satisfied
        
        // Update connection type
        if path.usesInterfaceType(.wifi) {
            connectionType = .wifi
        } else if path.usesInterfaceType(.cellular) {
            connectionType = .cellular
        } else if path.usesInterfaceType(.wiredEthernet) {
            connectionType = .ethernet
        } else if path.status == .satisfied {
            connectionType = .other
        } else {
            connectionType = .unknown
        }
        
        // Notify C++ engine selector if status changed
        if wasOnline != isOnline {
            statusCallback?(isOnline)
        }
        
        // Log connectivity changes for debugging
        print("ConnectivityMonitor: Status changed to \(isOnline ? "online" : "offline") via \(connectionType)")
    }
    
    /// Get current connectivity status
    func getCurrentStatus() -> (isOnline: Bool, connectionType: ConnectionType) {
        return (isOnline, connectionType)
    }
    
    /// Check if specific interface is available
    func isInterfaceAvailable(_ interfaceType: NWInterface.InterfaceType) -> Bool {
        return monitor.currentPath.usesInterfaceType(interfaceType)
    }
    
    /// Get detailed network information
    func getNetworkInfo() -> NetworkInfo {
        let path = monitor.currentPath
        return NetworkInfo(
            isOnline: path.status == .satisfied,
            connectionType: connectionType,
            availableInterfaces: getAvailableInterfaces(path: path),
            isExpensive: path.isExpensive,
            isConstrained: path.isConstrained
        )
    }
    
    /// Get available network interfaces
    private func getAvailableInterfaces(path: NWPath) -> [NWInterface.InterfaceType] {
        var interfaces: [NWInterface.InterfaceType] = []
        
        if path.usesInterfaceType(.wifi) {
            interfaces.append(.wifi)
        }
        if path.usesInterfaceType(.cellular) {
            interfaces.append(.cellular)
        }
        if path.usesInterfaceType(.wiredEthernet) {
            interfaces.append(.wiredEthernet)
        }
        if path.usesInterfaceType(.loopback) {
            interfaces.append(.loopback)
        }
        if path.usesInterfaceType(.other) {
            interfaces.append(.other)
        }
        
        return interfaces
    }
}

/// Network information structure
struct NetworkInfo {
    let isOnline: Bool
    let connectionType: ConnectivityMonitor.ConnectionType
    let availableInterfaces: [NWInterface.InterfaceType]
    let isExpensive: Bool
    let isConstrained: Bool
}

/// C++ bridge for connectivity monitoring
class ConnectivityBridge {
    private let monitor = ConnectivityMonitor()
    private var engineSelector: UnsafeMutableRawPointer?
    
    init() {
        // Set up callback to notify C++ engine selector
        monitor.setStatusCallback { [weak self] isOnline in
            self?.notifyEngineSelector(isOnline: isOnline)
        }
    }
    
    /// Set the C++ engine selector pointer
    func setEngineSelector(_ selector: UnsafeMutableRawPointer) {
        engineSelector = selector
    }
    
    /// Get current connectivity status
    func getCurrentStatus() -> Bool {
        return monitor.isOnline
    }
    
    /// Start monitoring
    func startMonitoring() {
        monitor.startMonitoring()
    }
    
    /// Stop monitoring
    func stopMonitoring() {
        monitor.stopMonitoring()
    }
    
    /// Notify C++ engine selector of status change
    private func notifyEngineSelector(isOnline: Bool) {
        guard let selector = engineSelector else { return }
        
        // Call C++ function to update connectivity status
        // This would be implemented in the C++ bridge
        updateConnectivityStatus(selector, isOnline)
    }
}

/// C function to update connectivity status in C++ engine selector
/// This function would be implemented in the C++ bridge
@_cdecl("updateConnectivityStatus")
func updateConnectivityStatus(_ selector: UnsafeMutableRawPointer, _ isOnline: Bool) {
    // Implementation would call the C++ EngineSelector::setConnectivityStatus
    // This is a placeholder for the actual C++ bridge implementation
    print("C++ Bridge: Updating connectivity status to \(isOnline ? "online" : "offline")")
}

File: android/ConnectivityMonitor.kt
package com.studyhive.core

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Build
import androidx.annotation.RequiresApi

/**
 * Android connectivity monitor using ConnectivityManager.NetworkCallback
 * Tracks network status and notifies the C++ engine selector
 */
class ConnectivityMonitor private constructor(private val context: Context) {
    
    private val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    private var networkCallback: ConnectivityManager.NetworkCallback? = null
    private var statusCallback: ((Boolean) -> Unit)? = null
    
    @Volatile
    private var isOnline: Boolean = false
    
    @Volatile
    private var connectionType: ConnectionType = ConnectionType.UNKNOWN
    
    enum class ConnectionType {
        UNKNOWN, WIFI, CELLULAR, ETHERNET, OTHER
    }
    
    companion object {
        @Volatile
        private var INSTANCE: ConnectivityMonitor? = null
        
        fun getInstance(context: Context): ConnectivityMonitor {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: ConnectivityMonitor(context.applicationContext).also { INSTANCE = it }
            }
        }
    }
    
    /**
     * Start monitoring network connectivity
     */
    fun startMonitoring() {
        if (networkCallback != null) {
            return // Already monitoring
        }
        
        networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                super.onAvailable(network)
                updateConnectionStatus()
            }
            
            override fun onLost(network: Network) {
                super.onLost(network)
                updateConnectionStatus()
            }
            
            override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
                super.onCapabilitiesChanged(network, networkCapabilities)
                updateConnectionStatus()
            }
        }
        
        val networkRequest = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()
        
        connectivityManager.registerNetworkCallback(networkRequest, networkCallback!!)
        
        // Get initial status
        updateConnectionStatus()
    }
    
    /**
     * Stop monitoring network connectivity
     */
    fun stopMonitoring() {
        networkCallback?.let { callback ->
            connectivityManager.unregisterNetworkCallback(callback)
            networkCallback = null
        }
    }
    
    /**
     * Set callback for connectivity status changes
     */
    fun setStatusCallback(callback: (Boolean) -> Unit) {
        statusCallback = callback
    }
    
    /**
     * Update connection status based on current network state
     */
    private fun updateConnectionStatus() {
        val wasOnline = isOnline
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val network = connectivityManager.activeNetwork
            val capabilities = connectivityManager.getNetworkCapabilities(network)
            
            isOnline = capabilities?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
            
            // Determine connection type
            connectionType = when {
                capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) == true -> ConnectionType.WIFI
                capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) == true -> ConnectionType.CELLULAR
                capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) == true -> ConnectionType.ETHERNET
                capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH) == true -> ConnectionType.OTHER
                else -> ConnectionType.UNKNOWN
            }
        } else {
            // Fallback for older Android versions
            @Suppress("DEPRECATION")
            val networkInfo = connectivityManager.activeNetworkInfo
            isOnline = networkInfo?.isConnected == true
            
            @Suppress("DEPRECATION")
            connectionType = when (networkInfo?.type) {
                ConnectivityManager.TYPE_WIFI -> ConnectionType.WIFI
                ConnectivityManager.TYPE_MOBILE -> ConnectionType.CELLULAR
                ConnectivityManager.TYPE_ETHERNET -> ConnectionType.ETHERNET
                else -> ConnectionType.UNKNOWN
            }
        }
        
        // Notify C++ engine selector if status changed
        if (wasOnline != isOnline) {
            statusCallback?.invoke(isOnline)
        }
        
        // Log connectivity changes for debugging
        android.util.Log.d("ConnectivityMonitor", "Status changed to ${if (isOnline) "online" else "offline"} via $connectionType")
    }
    
    /**
     * Get current connectivity status
     */
    fun getCurrentStatus(): Pair<Boolean, ConnectionType> {
        return Pair(isOnline, connectionType)
    }
    
    /**
     * Check if specific transport is available
     */
    @RequiresApi(Build.VERSION_CODES.M)
    fun isTransportAvailable(transport: Int): Boolean {
        val network = connectivityManager.activeNetwork
        val capabilities = connectivityManager.getNetworkCapabilities(network)
        return capabilities?.hasTransport(transport) == true
    }
    
    /**
     * Get detailed network information
     */
    fun getNetworkInfo(): NetworkInfo {
        return NetworkInfo(
            isOnline = isOnline,
            connectionType = connectionType,
            isExpensive = isConnectionExpensive(),
            isConstrained = isConnectionConstrained()
        )
    }
    
    /**
     * Check if connection is expensive (e.g., cellular data)
     */
    @RequiresApi(Build.VERSION_CODES.M)
    private fun isConnectionExpensive(): Boolean {
        val network = connectivityManager.activeNetwork
        val capabilities = connectivityManager.getNetworkCapabilities(network)
        return capabilities?.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED) != true
    }
    
    /**
     * Check if connection is constrained (e.g., low bandwidth)
     */
    @RequiresApi(Build.VERSION_CODES.M)
    private fun isConnectionConstrained(): Boolean {
        val network = connectivityManager.activeNetwork
        val capabilities = connectivityManager.getNetworkCapabilities(network)
        return capabilities?.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_RESTRICTED) != true
    }
}

/**
 * Network information data class
 */
data class NetworkInfo(
    val isOnline: Boolean,
    val connectionType: ConnectivityMonitor.ConnectionType,
    val isExpensive: Boolean,
    val isConstrained: Boolean
)

/**
 * C++ bridge for connectivity monitoring
 */
class ConnectivityBridge {
    private val monitor: ConnectivityMonitor
    private var engineSelector: Long = 0 // C++ pointer as Long
    
    constructor(context: Context) {
        monitor = ConnectivityMonitor.getInstance(context)
        
        // Set up callback to notify C++ engine selector
        monitor.setStatusCallback { isOnline ->
            notifyEngineSelector(isOnline)
        }
    }
    
    /**
     * Set the C++ engine selector pointer
     */
    fun setEngineSelector(selector: Long) {
        engineSelector = selector
    }
    
    /**
     * Get current connectivity status
     */
    fun getCurrentStatus(): Boolean {
        return monitor.getCurrentStatus().first
    }
    
    /**
     * Start monitoring
     */
    fun startMonitoring() {
        monitor.startMonitoring()
    }
    
    /**
     * Stop monitoring
     */
    fun stopMonitoring() {
        monitor.stopMonitoring()
    }
    
    /**
     * Notify C++ engine selector of status change
     */
    private fun notifyEngineSelector(isOnline: Boolean) {
        if (engineSelector != 0L) {
            // Call C++ function to update connectivity status
            // This would be implemented in the C++ bridge
            updateConnectivityStatus(engineSelector, isOnline)
        }
    }
}

/**
 * External C++ function to update connectivity status in engine selector
 * This function would be implemented in the C++ bridge
 */
external fun updateConnectivityStatus(selector: Long, isOnline: Boolean)

/**
 * Companion object for native library loading
 */
object ConnectivityBridgeNative {
    init {
        System.loadLibrary("studyhive_core")
    }
}
